src/infer_request.cc

```diff
@@ -1186,6 +1186,11 @@ InferenceRequest::Normalize()
       if (data_type == inference::DataType::TYPE_STRING) {
         RETURN_IF_ERROR(
             ValidateBytesInputs(input_id, input, &expected_byte_size));
+        // FIXME: -1 is used as a signal to skip total byte size validation for
+        // unhandled cases in ValidateBytesInputs.
+        if (expected_byte_size == -1) {
+          return Status::Success;
```

Comment:
I don't think we should return immediately incase more checks are appended to this function.
My concern is someone will add new lines without looking backwards. You should consider
```
bool skip_byte_size_check = (expected_byte_size == -1);

if (!skip_byte_size_check &&
    ((byte_size > INT_MAX) || (static_cast<int64_t>(byte_size) != expected_byte_size))) {
```

Reply:
Agree that you should not return early on "success". And further more, I think you should reorder the code here because `ValidateBytesInputs()` already does "total byte size validation":
```
if (byte_data_type) {
  RETURN_IF_ERROR(ValidateBytesInputs(...));
}
else { // non_byte_data
  // line 1179-1185
  // line 1195
  // line 1197-1205
}
```

Reply:
I also just realized that all these checks are inside of a for loop, so you probably intended to`continue` instead of `return` (neither was a good idea). It's prone to make mistakes which makes me feel like to refactor such a long function `Normalize`.

Reply:
> Agree that you should not return early on "success". And further more, I think you should reorder the code here because `ValidateBytesInputs()` already does "total byte size validation":
> 
> ```
> if (byte_data_type) {
>   RETURN_IF_ERROR(ValidateBytesInputs(...));
> }
> else { // non_byte_data
>   // line 1179-1185
>   // line 1195
>   // line 1197-1205
> }
> ```

`ValidateBytesInputs` asserts all buffers and all strings are visited. It doesn't check the size of the last string imo. I still think it's a good idea to compare expected_byte_size at one place regardless of input data type.

Reply:
Let me know if there is misalignment in my understanding:

"total byte size" in this function scope is `byte_size` which is set to `input.Data()->TotalByteSize()`.

In `ValidateBytesInputs()`, it iterates all elements in the data to get the actual byte size used in execution, and compare it against the buffer boundary (using buffer count and buffer byte size). Isn't the buffer boundary the reflection of `TotalByteSize()`? If iteration exactly ends at buffer boundary, then it implies the `expected size == got size`

Reply:
If iteration exactly ends at buffer boundary, it only means the number of bytes matches `kElementSizeIndicator`.

Reply:
"Iteration ends" refer to the point after checking all the elements, does that make sense?

Reply:
Gotcha. Here `byte_size` is comparing to itself (expected_byte_size) which is redundant.

Reply:
Dumb question: Is `input.Data()->TotalByteSize()` guarenteed to be accurate? I can see this check is useful when there is a size mismatch between `byte_size` and bytes get read from all buffers.

Reply:
>Is input.Data()->TotalByteSize() guaranteed to be correct?

This function should return the total size of the buffers.

> mismatch between `byte_size` and bytes get read from all buffers

From the current implementation they are the same thing, right? I think the comparison should be between buffer byte size (from `TotalByteSize()`) and byte size computed from other sources (i.e. shape * data byte size for non-byte tensor, and iterating elements for byte tensor).